### PR TITLE
AWS: Fix S3InputStream retry policy

### PR DESCRIPTION
**Pull request by edgarRd**
The retry policy for `S3InputStream` reads introduced in https://github.com/apache/iceberg/commit/c0d73f4ef5c16401bdfd62e1745faf2fbbf62177 (https://github.com/apache/iceberg/pull/10433) is not actually re-opening the stream on each retry attempt given that it uses `onFailure` which, as per the [documentation](https://failsafe.dev/javadoc/core/dev/failsafe/PolicyBuilder.html#onFailure-dev.failsafe.event.EventListener-), it is triggered after the whole (retry) policy has failed completely and was unable to produce a successful result, i.e. all retries have been exhausted and failed. With the existing implementation, once a stream fails with something like `SSLException` due to a socket connection reset, the retries will keep failing (due to not actually reopening the stream) and therefore this won't work. The `onFailure` call does not work either because by the time `openStream` is called, the retry policy failure has already been propagated, failing the read operation.
The tests also were not testing that the retry policy actually called `openStream(true)` to reset the stream, it only tested that the calls went through the retry policy a given number of times, generally always succeeding on the last one.

This PR fixes the problem by using `onRetry` instead, as per the [documentation](https://failsafe.dev/javadoc/core/dev/failsafe/RetryPolicyBuilder.html#onRetry-dev.failsafe.event.EventListener-), to re-open the stream right before we attempt the read again. Also, it fixes the tests by actually checking that the stream is reset (re-opened).

### CODE REVIEW

**Comment by edgarRd** [comment]
@amogh-jahagirdar @jackye1995 @danielcweeks PTAL. Thanks!

**Comment by edgarRd** [comment]
cc @Parth-Brahmbhatt

**Review by Parth-Brahmbhatt — APPROVED** [review]

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 1]
```diff
@@ -66,10 +70,14 @@ class S3InputStream extends SeekableInputStream implements RangeReadable {
...
-                  SSLException.class, SocketTimeoutException.class, SocketException.class))
-          .onFailure(failure -> openStream(true))
+          .handle(RETRYABLE_EXCEPTIONS)
+          .onRetry(
+              e -> {
+                LOG.warn(
+                    "Retrying read from S3, reopening stream (attempt {})", e.getAttemptCount());
+                resetForRetry();
+              })
+          .onFailure(e -> LOG.error("Failed to read from S3 after retry policy", e.getException()))
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:80]: "Failed to read from S3 input stream after exhausting all retries"

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestFlakyS3InputStream.java [hunk 2]
```diff
@@ -49,10 +51,29 @@
...
+  }
+
+  @Override
+  S3InputStream newInputStream(S3Client s3Client, S3URI uri) {
+    return new S3InputStream(s3Client, uri) {
+      @Override
+      void resetForRetry() throws IOException {
+        resetForRetryCounter.incrementAndGet();
+        super.resetForRetry();
+      }
```
Comment (amogh-jahagirdar) [inline aws/src/test/java/org/apache/iceberg/aws/s3/TestFlakyS3InputStream.java:68]: Thanks for fixing this, this was crucial for verifying the retry behavior actually resets the input stream. Seems like we were just getting lucky on the tests before.

**Review by amogh-jahagirdar — COMMENTED** [review]
Thanks for fixing this @edgarRd , this was a miss on my part on the original PR. Yeah we need to be using `onRetry` for triggering the input stream reset, not `onFailure`. Had some minor comments on the log message, would be great to fix so we can get this correctly into 1.7

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestFlakyS3InputStream.java (hunk 2 above)
Comment (edgarRd) [reply aws/src/test/java/org/apache/iceberg/aws/s3/TestFlakyS3InputStream.java:68]: Yes. That's correct.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 1 above)
Comment (edgarRd) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java]: Done.

**Comment by edgarRd** [comment]
@amogh-jahagirdar I've address the comment. PTAL - Thanks.

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks @edgarRd, really appreciate the fix here! I'll keep it open for a little bit in case anyone else had any comments.

**Review by singhpk234 — APPROVED** [review]
LGTM as well, Thanks @edgarRd !

**Comment by amogh-jahagirdar** [comment]
Thanks @edgarRd, I'll go ahead and merge. Thank you @singhpk234 for reviewing!

**Comment by edgarRd** [comment]
Thanks for the reviews, @amogh-jahagirdar @singhpk234 and @Parth-Brahmbhatt !

**Pull request merged**